src/ansi-c/c_nondet_symbol_factory.cpp

```diff
@@ -0,0 +1,419 @@
+  aux_symbol.is_static_lifetime=static_lifetime;
+
+  exprt object=aux_symbol.symbol_expr();
+  exprt aoe=address_of_exprt(object);
+
+  const typet &allocate_type_resolved=ns.follow(allocate_type);
+  const typet &target_type=ns.follow(target_expr.type().subtype());
+  bool cast_needed=allocate_type_resolved!=target_type;
+  if(cast_needed)
+  {
+    aoe=typecast_exprt(aoe, target_expr.type());
+  }
+
+  code_assignt assign(target_expr, aoe);
+  assign.add_source_location()=loc;
+  code.add(assign);
+
+  input_recorder.record_input(aux_symbol.base_name, object);
+
+  return aoe;
+}
+
+/*******************************************************************\
+
+Function: symbol_factoryt::gen_nondet_init
+
+ Inputs:
+  init_code - The code block to add code to
+  expr - The expression which we are generating a non-determinate value for
+
+ Outputs:
+
+ Purpose: Creates a nondet for expr, including calling itself recursively to
+          make appropriate symbols to point to if expr is a pointer.
+
+\*******************************************************************/
+
+void symbol_factoryt::gen_nondet_init(
+  code_blockt& init_code,
+  const exprt &expr)
+{
+  const typet &type=ns.follow(expr.type());
+
+  if(type.id()==ID_pointer)
+  {
+    // dereferenced type
+    const pointer_typet &pointer_type=to_pointer_type(type);
+    const typet &subtype=ns.follow(pointer_type.subtype());
+
+    code_blockt non_null_inst;
+
+    // Note, the allocated object needs to be static. E.g., with a pointer to a
+    // pointer to a pointer temporary variables are created in else clauses,
+    // and thus go out of scope before the function is called.
+    exprt allocated=
+      allocate_object(non_null_inst, expr, subtype, true);
+
+    exprt init_expr;
+    if(allocated.id()==ID_address_of)
+    {
+      init_expr=allocated.op0();
+    }
+    else
+    {
+      init_expr=dereference_exprt(allocated, allocated.type().subtype());
+    }
+    gen_nondet_init(non_null_inst, init_expr);
+
+    if(assume_non_null)
+    {
+      // Add the following code to init_code:
+      // <expr> = <aoe>;
+      concatenate_code_blocks(init_code, non_null_inst);
+    }
+    else
+    {
+      // Add the following code to init_code:
+      //           IF !(NONDET(_Bool) == FALSE) THEN GOTO <label1>
+      //           <expr> = <null pointer>
+      //           GOTO 2
+      // <label1>: <expr> = &tmp$<temporary_counter>;
+      //           <code from recursive call to gen_nondet_init() with
+      //             tmp$<temporary_counter>>
+      // And the next line is labelled label2
+      auto set_null_inst=code_assignt(expr, null_pointer_exprt(pointer_type));
+      set_null_inst.add_source_location()=loc;
+
+      const exprt nondet_bool=c_get_nondet_bool(c_bool_typet(1));
+      const exprt null_return=from_integer(0, c_bool_typet(1));
+
+      code_ifthenelset null_check;
+      null_check.cond()=equal_exprt(nondet_bool, null_return);
```

Comment:
Actually that generated an error when running without --show-goto-functions. I ran on pointer-function-parameters/main.c with the flags --function fun --cover branch and I got this output:
```
Passing problem to propositional reduction
converting SSA
######### lhs: symbol
  * type: bool
  * identifier: goto_symex::\guard#1
  * expression: symbol
      * type: bool
      * identifier: goto_symex::\guard
  * L2: 1
  * L1_object_identifier: goto_symex::\guard
  * #SSA_symbol: 1
######### rhs: typecast
  * type: c_bool
      * width: 1
  0: nondet_symbol
      * type: bool
      * identifier: symex::nondet0
      * #source_location: nil
equality without matching types
/home/owen/workspace/cbmc/src/cbmc/cbmc exited with code 10
```